File: solution/day07/src/main/java/ci/functional/Pipeline.java
package ci.functional;

import ci.dependencies.Config;
import ci.dependencies.Emailer;
import ci.dependencies.Logger;
import ci.dependencies.Project;
import io.vavr.control.Option;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.With;

import static io.vavr.control.Option.some;

@AllArgsConstructor
public class Pipeline {
    private final Config config;
    private final Emailer emailer;
    private final Logger log;
    private static final String SUCCESS = "success";

    public void run(Project project) {
        createProjectContext(project)
                .flatMap(this::runTests)
                .flatMap(this::deploy);
    }

    private Option<PipelineContext> createProjectContext(Project project) {
        return some(new PipelineContext(project))
                .peek(context -> {
                    if (!context.isHasTests()) log.info("No tests");
                });
    }

    private Option<PipelineContext> runTests(PipelineContext context) {
        return context.hasTests
                ? runTestsSafely(context)
                : some(context);
    }

    private Option<PipelineContext> runTestsSafely(PipelineContext context) {
        return some(context.getProject().runTests())
                .map(testsResult -> context.withTestsRanSuccessfully(testsResult.equals(SUCCESS)))
                .peek(this::logTestResult);
    }

    private void logTestResult(PipelineContext context) {
        if (!context.isTestsRanSuccessfully()) {
            log.error("Tests failed");
            sendEmail("Tests failed");
            return;
        }
        log.info("Tests passed");
    }

    private Option<PipelineContext> deploy(PipelineContext context) {
        return context.mustRunDeployment()
                ? deploySafely(context)
                : some(context);
    }

    private Option<PipelineContext> deploySafely(PipelineContext context) {
        return some(context.getProject().deploy())
                .map(deploymentResult -> context.withDeployedSuccessfully(deploymentResult.equals(SUCCESS)))
                .peek(this::logDeploymentResult);
    }

    private void logDeploymentResult(PipelineContext context) {
        if (!context.isDeployedSuccessfully()) {
            log.error("Deployment failed");
            sendEmail("Deployment failed");
            return;
        }
        log.info("Deployment successful");
        sendEmail("Deployment completed successfully");
    }

    private void sendEmail(String text) {
        if (!config.sendEmailSummary()) {
            log.info("Email disabled");
            return;
        }
        log.info("Sending email");
        emailer.send(text);
    }

    @Getter
    @AllArgsConstructor
    @With
    private static class PipelineContext {
        private final boolean hasTests;
        private final Project project;
        private boolean testsRanSuccessfully;
        private boolean deployedSuccessfully;

        public PipelineContext(Project project) {
            this.project = project;
            this.hasTests = project.hasTests();
        }

        public boolean mustRunDeployment() {
            return testsRanSuccessfully || !hasTests;
        }
    }
}

File: solution/ts/day07/src/pipeline.ts
import {Config, Emailer, Logger} from "./ci";
import {Project} from "./project";
import * as O from 'fp-ts/Option';
import {Option, some} from 'fp-ts/Option';
import {pipe} from "fp-ts/function";

const success = 'success';

export class Pipeline {
    constructor(private config: Config, private emailSender: Emailer, private log: Logger) {
    }

    run(project: Project): void {
        pipe(
            this.createProjectContext(project),
            O.chain(context => this.runTests(context)),
            O.chain(context => this.deploy(context))
        );
    }

    private createProjectContext(project: Project): Option<PipelineContext> {
        if (!project.hasTests()) {
            this.log.info('No tests');
        }
        return some(new PipelineContext(project));
    }

    private runTests = (context: PipelineContext): Option<PipelineContext> =>
        context.hasTests
            ? this.runTestsSafely(context)
            : some(context);

    private runTestsSafely = (context: PipelineContext): Option<PipelineContext> => some(
        this.logTestResult(
            context.withTestsResults(context.project.runTests() === success)
        )
    );

    private logTestResult(context: PipelineContext): PipelineContext {
        if (!context.testsRanSuccessfully) {
            this.log.error('Tests failed');
            this.sendEmail('Tests failed');
        } else {
            this.log.info('Tests passed');
        }
        return context;
    }

    private deploy = (context: PipelineContext): Option<PipelineContext> =>
        context.mustRunDeployment()
            ? this.deploySafely(context)
            : some(context);

    private deploySafely = (context: PipelineContext): Option<PipelineContext> => some(
        this.logDeploymentResult(
            context.withDeploymentStatus(context.project.deploy() === success)
        )
    );

    private logDeploymentResult(context: PipelineContext): PipelineContext {
        if (!context.deployedSuccessfully) {
            this.log.error('Deployment failed');
            this.sendEmail('Deployment failed');
        } else {
            this.log.info('Deployment successful');
            this.sendEmail('Deployment completed successfully');
        }
        return context;
    }

    private sendEmail(text: string): void {
        if (!this.config.sendEmailSummary()) {
            this.log.info('Email disabled');
            return;
        }
        this.log.info('Sending email');
        this.emailSender.send(text);
    }
}

class PipelineContext {
    public hasTests = this.project.hasTests();
    public testsRanSuccessfully = false;
    public deployedSuccessfully = false;

    constructor(public readonly project: Project) {
    }

    mustRunDeployment(): boolean {
        return this.testsRanSuccessfully || !this.hasTests;
    }

    withTestsResults(testsResult: boolean): PipelineContext {
        this.testsRanSuccessfully = testsResult
        return this
    }

    withDeploymentStatus(deploymentStatus: boolean): PipelineContext {
        this.deployedSuccessfully = deploymentStatus
        return this
    }
}